File: packages/emitter/testcases/real-world/functional/functional.ts
// Option type (Maybe monad)
export type Option<T> = Some<T> | None;

export class Some<T> {
  readonly _tag = "Some";
  constructor(readonly value: T) {}
}

export class None {
  readonly _tag = "None";
}

export function some<T>(value: T): Option<T> {
  return new Some(value);
}

export function none<T>(): Option<T> {
  return new None();
}

export function isSome<T>(option: Option<T>): option is Some<T> {
  return (option as Some<T>)._tag === "Some";
}

export function isNone<T>(option: Option<T>): option is None {
  return (option as None)._tag === "None";
}

export function getOrElse<T>(option: Option<T>, defaultValue: T): T {
  if (isSome(option)) {
    return option.value;
  }
  return defaultValue;
}

// Pure functions for array operations
export function pipe<A, B>(value: A, fn: (a: A) => B): B {
  return fn(value);
}

export function compose<A, B, C>(
  f: (b: B) => C,
  g: (a: A) => B
): (a: A) => C {
  return (a: A) => f(g(a));
}

export function curry<A, B, C>(
  fn: (a: A, b: B) => C
): (a: A) => (b: B) => C {
  return (a: A) => (b: B) => fn(a, b);
}

export function partial<A, B, C>(
  fn: (a: A, b: B) => C,
  a: A
): (b: B) => C {
  return (b: B) => fn(a, b);
}

// Higher-order functions
export function memoize<T extends any[], R>(
  fn: (...args: T) => R
): (...args: T) => R {
  const cache = new Map<string, R>();

  return (...args: T): R => {
    const key = JSON.stringify(args);
    if (cache.has(key)) {
      return cache.get(key)!;
    }

    const result = fn(...args);
    cache.set(key, result);
    return result;
  };
}

export function debounce<T extends any[]>(
  fn: (...args: T) => void,
  delayMs: number
): (...args: T) => void {
  let timeoutId: any;

  return (...args: T): void => {
    if (timeoutId !== undefined) {
      clearTimeout(timeoutId);
    }

    timeoutId = setTimeout(() => {
      fn(...args);
    }, delayMs);
  };
}

// Lazy evaluation
export class Lazy<T> {
  private value: T | undefined;
  private computed: boolean = false;

  constructor(private computer: () => T) {}

  get(): T {
    if (!this.computed) {
      this.value = this.computer();
      this.computed = true;
    }
    return this.value!;
  }

  map<U>(mapper: (value: T) => U): Lazy<U> {
    return new Lazy(() => mapper(this.get()));
  }
}

// Immutable operations
export function assoc<T, K extends keyof T>(
  obj: T,
  key: K,
  value: T[K]
): T {
  return { ...obj, [key]: value };
}

export function dissoc<T, K extends keyof T>(
  obj: T,
  key: K
): Omit<T, K> {
  const { [key]: _, ...rest } = obj;
  return rest;
}

File: packages/emitter/testcases/real-world/functional/functional.cs

using Tsonic.Runtime;
using System;
using System.Collections.Generic;

namespace TestCases.realworld
{
    public class Some<T>
    {
        public readonly object _tag = "Some";

        public Some(T value)
            {

            }
    }
    public class None
    {
        public readonly object _tag = "None";
    }
    public class Lazy<T>
    {
        private Func<T> computer;

        private T? value;

        private bool computed = false;

        public Lazy(Func<T> computer)
            {
            this.computer = computer;
            }

        public T get()
            {
            if (!this.computed)
                {
                this.value = this.computer();
                this.computed = true;
                }
            return this.value!;
            }

        public Lazy<U> map<U>(Func<T, U> mapper)
            {
            return new Lazy(() => mapper(this.get()));
            }
    }

    public static class functional
    {
        // type Option = Union<Some<T>, None>

        public static Option<T> some<T>(T value)
            {
            return new Some(value);
            }

        public static Option<T> none<T>()
            {
            return new None();
            }

        public static dynamic isSome<T>(Option<T> option)
            {
            return option._tag == "Some";
            }

        public static dynamic isNone<T>(Option<T> option)
            {
            return option._tag == "None";
            }

        public static T getOrElse<T>(Option<T> option, T defaultValue)
            {
            if (isSome(option))
                {
                return option.value;
                }
            return defaultValue;
            }

        public static B pipe<A, B>(A value, Func<A, B> fn)
            {
            return fn(value);
            }

        public static Func<A, C> compose<A, B, C>(Func<B, C> f, Func<A, B> g)
            {
            return (a) => f(g(a));
            }

        public static Func<A, Func<B, C>> curry<A, B, C>(Func<A, B, C> fn)
            {
            return (a) => (b) => fn(a, b);
            }

        public static Func<B, C> partial<A, B, C>(Func<A, B, C> fn, A a)
            {
            return (b) => fn(a, b);
            }

        public static Func<T, R> memoize<T, R>(Func<T, R> fn)
            where T : List<dynamic>
            {
            var cache = new Map<string, R>();
            return (args) =>
            {
            var key = Tsonic.Runtime.JSON.stringify(args);
            if (cache.has(key))
                {
                return cache.get(key)!;
                }
            var result = fn(params args);
            cache.set(key, result);
            return result;
            };
            }

        public static Action<T> debounce<T>(Action<T> fn, double delayMs)
            where T : List<dynamic>
            {
            dynamic timeoutId;
            return (args) =>
            {
            if (timeoutId != default)
                {
                clearTimeout(timeoutId);
                }
            timeoutId = setTimeout(() =>
            {
            fn(params args);
            }, delayMs);
            };
            }

        public static T assoc<T, K>(T obj, K key, dynamic value)
            where K : dynamic
            {
            return new { /* ...spread */, /* computed */ = value };
            }

        public static Omit<T, K> dissoc<T, K>(T obj, K key)
            where K : dynamic
            {
            var /* destructuring */ = obj;
            return rest;
            }
    }
}